### PR TITLE
Add functions to read/write a_ℓm and C_ℓ from/to FITS files

### PR DESCRIPTION
**Pull request by LeeoBianchi**
Added writeAlmToFITS, writeClToFITS, readClfromFITS, brief markdown documentation chapter for the Cl-related functions.

### CODE REVIEW

**Review by ziotom78 — COMMENTED** [review]
Hi @LeoBianchi, might you please remove from the commit the files in the hidden directory `docs/src/.ipynb_checkpoints`? In this way the PR looks cleaner. Thanks!

File: src/Cl.jl [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    numOfRows = CFITSIO.fits_get_num_rows(f)
+
+    Cl = Vector{T}(undef, numOfRows)
+
+    CFITSIO.fits_read_col(f, 2, 1, 1, Cl)
+
+    return Cl
+end
+
+function readClFromFITS(fileName, t::Type{T}) where {T <: Real}
```
Comment (ziotom78) [inline src/Cl.jl:16]: Same here, a `colnum = 2` keyword would be nice.

File: src/alm.jl [hunk 2]
```diff
@@ -177,3 +217,78 @@ function gaussbeam(fwhm::T, lmax::Int; pol=false) where T
     end
     return Bl
 end
+
+###########################################################################
+
+"""
+Multiply an a_lm by a vector b_l representing an l-dependendt function.
```
Comment (ziotom78) [inline src/alm.jl:224]: Typo: `dependendt` → `dependent`

File: src/alm.jl [hunk 3]
```diff
@@ -177,3 +217,78 @@ function gaussbeam(fwhm::T, lmax::Int; pol=false) where T
...
+            i = almIndex(alms, l, m)
+            alm_new.alm[i] = alms.alm[i]*f
+        end
+    end
+    return alm_new
+end
+
+
+"""
+Multiply IN-PLACE an a_lm by a vector b_l representing an l-dependendt function.
```
Comment (ziotom78) [inline src/alm.jl:263]: Same typo as above

**Review by ziotom78 — CHANGES_REQUESTED** [review]
Thanks @LeeoBianchi , nice PR! I added a few comments here and there, may you fix them?

File: src/Cl.jl [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+#ARGUMENTS:
+- `dl::Vector{T}` : Array of Cℓ components
+- `lmin::Int` : minimum l in the representation of the Dℓ power spectrum
+
+#RETURNS:
+- `Vector{T}` : Array of Dℓ power spectrum components
+"""
+
+function cl2dl(cl, lmin)
```
[the remark this answers is not in the document]
Comment (LeeoBianchi) [reply src/Cl.jl:106]: Ops, yes it did slip!

File: src/alm.jl [hunk 5]
```diff
@@ -177,3 +217,78 @@ function gaussbeam(fwhm::T, lmax::Int; pol=false) where T
...
+    return alm_new
+end
+
+
+"""
+Multiply IN-PLACE an a_lm by a vector b_l representing an l-dependent function.
+
+# ARGUMENTS
+- `alms::Alm{Complex{T}}`: The array representing the spherical harmonics coefficients
+- `fl::Vector{T}`: The array giving the factor f_l by which to multiply a_lm
```
Comment (ziotom78) [inline src/alm.jl:267]: `Vector` should be `AbstractVector`

File: src/alm.jl [hunk 6]
```diff
@@ -177,3 +217,78 @@ function gaussbeam(fwhm::T, lmax::Int; pol=false) where T
...
 end
+
+###########################################################################
+
+"""
+Multiply an a_lm by a vector b_l representing an l-dependent function.
+
+# ARGUMENTS
+- `alms::Alm{Complex{T}}`: The array representing the spherical harmonics coefficients
+- `fl::Vector{T}`: The array giving the factor f_l by which to multiply a_lm
```
Comment (ziotom78) [inline src/alm.jl:228]: `Vector` should be `AbstractVector`

File: src/Cl.jl [hunk 7]
```diff
@@ -0,0 +1,114 @@
+"""
+    readClFromFITS{T <: Real}(f::CFITSIO.FITSFile, t::Type{T}) -> Vector{T}
+    readClFromFITS{T <: Real}(fileName::String, t::Type{T}) -> Vector{T}
+Read a set of C_ℓ coefficients from a FITS file.
```
Comment (ziotom78) [inline src/Cl.jl:4]: Update the docstring to include `col_num`

File: src/alm.jl [hunk 8]
```diff
@@ -108,6 +112,90 @@ function readAlmFromFITS(fileName, t::Type{T}) where {T <: Complex}
...
+    end
+    idx = Vector{Int}()
+    for m = 0:mmax
+        for l = m:lmax
+            i = l^2 + l + m + 1
+            append!(idx, i)
+        end
+    end
+    return idx
+end
```
Comment (ziotom78) [inline src/alm.jl:141]: This code is quite inefficient, as the array needs to grow progressively every time `append!` is invoked. Also, it's better to avoid the parameter `alm` to be passed, as the only meaningful inputs for this function are `lmax` and `mmax`.

The following code should work better:

```julia
# No need to specify the type for `lmax` and `mmax`
function almExplicitIndex2(lmax, mmax) where {T <: Number}
    # Step 1: count the number of elements in the output
    count = 0
    for m = 0:mmax
        (lmax ≥ m) && (count += (lmax - m) + 1)
    end

    # Step 2: allocate the vector for the output in one batch
    idx = Vector{Int}(undef, count)

    # Step 3: fill the output
    i = 1
    for m = 0:mmax
        for l = m:lmax
            idx[i] = l^2 + l + m + 1
            i += 1
        end
    end

    idx  # In Julia it's preferable to avoid "return" at the end of functions
end
```

I did a few measurements and I see a ~10 time increase in speed and a significant reduction in allocated memory.

Original implementation:

```
julia> using BenchmarkTools
julia> @benchmark almExplicitIndex(1000, 3000)
BenchmarkTools.Trial: 935 samples with 1 evaluation.
 Range (min … max):  4.547 ms …   6.836 ms  ┊ GC (min … max): 0.00% … 21.34%
 Time  (median):     5.435 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):   5.308 ms ± 631.319 μs  ┊ GC (mean ± σ):  1.56% ±  5.03%

   ▇█▄                        ▅▃▁                              
  ▇████▄▃▂▂▂▁▂▂▂▂▄▄▄▃▂▂▂▃▅▃▃▅████▆▅▃▃▃▂▂▂▂▂▂▁▂▂▁▂▁▁▄▄▃▃▃▃▄▃▄▃ ▃
  4.55 ms         Histogram: frequency by time        6.69 ms <

 Memory estimate: 5.23 MiB, allocs estimate: 13.
```

New implementation:

```
julia> @benchmark almExplicitIndex2(1000, 3000)
BenchmarkTools.Trial: 8343 samples with 1 evaluation.
 Range (min … max):  509.585 μs …   1.947 ms  ┊ GC (min … max): 0.00% … 29.00%
 Time  (median):     544.338 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   593.617 μs ± 154.445 μs  ┊ GC (mean ± σ):  7.20% ± 13.30%

  ▃█▆▇▆▆▅▄▃▂▁                                        ▁▁▂▂▂▂▁▁▁  ▂
  ████████████▇▇▅▄▇▆▅▅▃▃▅▄▄▅▃▁▃▁▄▄▁▃▄▅▄▅▅▆▄▅▅▅▄▃▄▄▄▅▅██████████ █
  510 μs        Histogram: log(frequency) by time       1.11 ms <

 Memory estimate: 3.83 MiB, allocs estimate: 2.

File: src/alm.jl [hunk 9]
```diff
@@ -108,6 +112,90 @@ function readAlmFromFITS(fileName, t::Type{T}) where {T <: Complex}
...
+
+    idx = Vector{Int}()
+    for m = 0:mmax
+        for l = m:lmax
+            i = l^2 + l + m + 1
+            append!(idx, i)
+        end
+    end
+    return idx
+end
```
Comment (ziotom78) [inline src/alm.jl:156]: Not sure why we need two implementations of `almExplicitIndex` like the ones above. Better to define

```julia
almExplicitIndex(alm::Alm{T}) where {T} = almExplicitIndex(alm.lmax, alm.mmax)
```

to rely on the function that I provided in my previous comment.

File: src/alm.jl (hunk 8 above)
Comment (LeeoBianchi) [reply src/alm.jl]: Thanks for the hint! I've fixed it with a new commit.

File: .gitignore [hunk 10]
```diff
@@ -1,3 +1,6 @@
 *~
 docs/build
 Manifest.toml
+src/Healpix.jl
+src/sphtfunc.jl
+test/test_sphtfunc.jl
```
Comment (ziotom78) [inline .gitignore:6]: Why were these files added to `.gitignore`?

Comment (LeeoBianchi) [reply .gitignore]: I was worried of pushing wrong changes while working on the other functions for the next PR. It was just temporary because I then created new files locally for that. I was sure I re-changed the `.gitignore` but I probably forgot to push it. It should be fixed now.

**Comment by LeeoBianchi** [comment]
From the test log it looked like a bracket was missing in alm.jl, it should work now.
